### ISSUE TITLE
Allow prediction_length of different length

### ISSUE THREAD

**Issue by StatMixedML** [awslabs/gluon-ts#688]
Currently, *prediction_length* is a global parameter that produces forecasts for all series in the test data set with the same length of *prediction_length*. This implicitly assumes that the number of rows for each time series in the test set must have the same number.

However, to be more flexible, it would be great if we would allow *prediction_length* to be a list where the user can specify an individual prediction horizon for each series, i.e., *prediction_length  = [2, 12, 4, 8]*.

**Comment by lostella** [comment]
> This implicitly assumes that the number of rows for each time series in the test set must have the same number.

I’m not sure I understand: what do you mean by “rows”?

**Comment by StatMixedML** [comment]
@lostella What I mean is that if, e.g.,  *prediction_length = 24*, the test set is ~ *train_set + prediction_length* so that for each time series, we add *prediction_length* rows to the train set on which the model then provides forecasts, right?

So instead of  *prediction_length* being the same for all test time series, we allow *prediction_length* to vary.

**Comment by kaleming** [comment]
Maybe it is related to this (#518)

**Comment by lostella** [comment]
I think it might be hard to enforce such an interface on all models: trainable models (like neural network based ones) often have a hard-coded prediction_length. That is to say, they consist exactly of the parameters that allow you to compute (some representation of) a forecast of length prediction_length given past data. In some cases this may be possible: the naive prediction, that uses the latest observation as forecast, can be extended for as many time steps you want. However, allowing such an interface and then having some model to support it and some others not, could turn out to be a bit messy in my opinion.

**Comment by lostella** [comment]
I'm replying to your comment in #519 here: it looks like that discussion is related to this one and off topic there.

> Assume that the prediction length = 18 months. However, most of the time series in the training set have length around 6-8 months, some longer

> So you`d suggest NOT TO PAD the training data and to set _context_length = min(train_length)_? Or how would you tackle the problem of short and irregularly spaced time series? Sorry for the confusion. Also, how can I set _pick_incomplete_ in the code?

Having time series shorter than prediction_length is a fundamental issue I believe: if you want to train a DeepAR model to predict 18 data points, you should provide it with training time series longer than 18 data points. This is regardless of the context_length: if data prior to the prediction range is not enough for some time series, the model will pad that automatically and learn to ignore those fake observations.

If some time series are shorter than prediction_length, then you can either pad them manually (I would discourage this) or exclude them from the training set.

By "irregularly spaced", I'm assuming you mean that time series don't all begin/end at the same points in time.

> Also, how can I set pick_incomplete in the code?

There's a constructor option in DeepAREstimator for that.

**Comment by mbohlkeschneider** [comment]
I agree with @lostella: Technically, that would be possible, I think, but I think it is hard to justify the extra complexity (I can imagine a lot of things that can go wrong with that).

**Comment by StatMixedML** [comment]
> There's a constructor option in DeepAREstimator for that.

@lostella May I ask you to show how to exactly use it. I cannot find it ?DeepAREstimator.

Thanks!

**Comment by lostella** [comment]
@StatMixedML Ah! My bad, sorry! It appears to be exposed in other estimators, but not in DeepAR. You can open a PR to include it, if you want. You can see how to option operates e.g. in DeepVAR here:

https://github.com/awslabs/gluon-ts/blob/52f7aac6e0efd090b90063a873d35ec5a2b8a1e6/src/gluonts/model/deepvar/_estimator.py#L224

In particular, it triggers the option in the `InstanceSplitter`, which is here:

https://github.com/awslabs/gluon-ts/blob/52f7aac6e0efd090b90063a873d35ec5a2b8a1e6/src/gluonts/transform/split.py#L128
